File: compress_images.py
#!/usr/bin/env python3
"""
=============================================================
  Script nén ảnh cho Wedding Website
  Tác giả: Antigravity
  Mô tả: Tự động tìm và nén tất cả ảnh JPG/PNG trong thư mục
         images/, lưu bản gốc vào images/_backup/
=============================================================

Cách dùng:
  1. Cài Pillow: pip install Pillow
  2. Chạy:       python compress_images.py

Tùy chỉnh:
  - JPEG_QUALITY  : Chất lượng ảnh JPG (1-95). Mặc định 75 (đẹp + nhẹ).
  - MAX_WIDTH     : Chiều rộng tối đa (px). Ảnh lớn hơn sẽ bị thu nhỏ.
  - MAX_HEIGHT    : Chiều cao tối đa (px).
  - BACKUP        : True = giữ bản gốc trong images/_backup/
=============================================================
"""

import os
import shutil
from pathlib import Path
from PIL import Image

# ─────────────────────────────────────────────
#  CẤU HÌNH – Chỉnh tại đây nếu cần
# ─────────────────────────────────────────────
IMAGES_DIR   = Path(__file__).parent / "images"   # Thư mục ảnh
JPEG_QUALITY = 75        # Chất lượng JPG (1-95). 75 = cân bằng đẹp & nhẹ
PNG_OPTIMIZE = True      # True = tối ưu PNG (không mất chất lượng)
MAX_WIDTH    = 1920      # Chiều rộng tối đa (px)
MAX_HEIGHT   = 1920      # Chiều cao tối đa (px)
BACKUP       = True      # True = lưu bản gốc vào _backup/
SKIP_DIRS    = {"_backup"}  # Bỏ qua các thư mục này
# ─────────────────────────────────────────────

SUPPORTED = {".jpg", ".jpeg", ".png", ".JPG", ".JPEG", ".PNG"}


def format_size(size_bytes: int) -> str:
    """Định dạng dung lượng file cho dễ đọc."""
    if size_bytes < 1024:
        return f"{size_bytes} B"
    elif size_bytes < 1024 ** 2:
        return f"{size_bytes / 1024:.1f} KB"
    else:
        return f"{size_bytes / 1024 ** 2:.2f} MB"


def backup_file(src: Path, backup_root: Path) -> None:
    """Sao lưu file gốc vào thư mục _backup giữ nguyên cấu trúc."""
    rel = src.relative_to(IMAGES_DIR)
    dest = backup_root / rel
    dest.parent.mkdir(parents=True, exist_ok=True)
    if not dest.exists():
        shutil.copy2(src, dest)


def compress_image(path: Path, backup_root: Path) -> tuple[int, int]:
    """
    Nén một file ảnh. Trả về (size_before, size_after).
    """
    size_before = path.stat().st_size

    if BACKUP:
        backup_file(path, backup_root)

    try:
        with Image.open(path) as img:
            # Chuyển RGBA/P → RGB cho JPG
            ext = path.suffix.lower()
            if ext in {".jpg", ".jpeg"} and img.mode in ("RGBA", "P", "LA"):
                img = img.convert("RGB")

            # Thu nhỏ nếu quá lớn (giữ tỉ lệ)
            w, h = img.size
            if w > MAX_WIDTH or h > MAX_HEIGHT:
                img.thumbnail((MAX_WIDTH, MAX_HEIGHT), Image.LANCZOS)

            # Lưu với nén
            if ext in {".jpg", ".jpeg"}:
                img.save(path, "JPEG", quality=JPEG_QUALITY, optimize=True)
            elif ext == ".png":
                img.save(path, "PNG", optimize=PNG_OPTIMIZE)

    except Exception as e:
        print(f"  ⚠️  Lỗi khi xử lý {path.name}: {e}")
        return size_before, size_before  # Không thay đổi

    size_after = path.stat().st_size
    return size_before, size_after


def find_images(root: Path) -> list[Path]:
    """Tìm tất cả ảnh được hỗ trợ, bỏ qua SKIP_DIRS."""
    result = []
    for item in root.rglob("*"):
        if item.is_file() and item.suffix in SUPPORTED:
            # Bỏ qua nếu nằm trong thư mục cần skip
            if any(skip in item.parts for skip in SKIP_DIRS):
                continue
            result.append(item)
    return sorted(result)


def main():
    print("=" * 60)
    print("  🖼️   SCRIPT NÉN ẢNH – WEDDING WEBSITE")
    print("=" * 60)

    if not IMAGES_DIR.exists():
        print(f"❌ Không tìm thấy thư mục: {IMAGES_DIR}")
        return

    backup_root = IMAGES_DIR / "_backup"
    images = find_images(IMAGES_DIR)

    if not images:
        print("⚠️  Không tìm thấy ảnh nào.")
        return

    print(f"\n📁 Thư mục ảnh : {IMAGES_DIR}")
    print(f"🔢 Tổng số ảnh  : {len(images)} file")
    print(f"🎯 Chất lượng  : JPG={JPEG_QUALITY}%, Max={MAX_WIDTH}x{MAX_HEIGHT}px")
    if BACKUP:
        print(f"💾 Backup       : {backup_root}")
    print()
    print("-" * 60)

    total_before = 0
    total_after  = 0

    for i, img_path in enumerate(images, 1):
        rel = img_path.relative_to(IMAGES_DIR)
        b, a = compress_image(img_path, backup_root)

        saved = b - a
        pct   = (saved / b * 100) if b > 0 else 0
        arrow = "✅" if saved > 0 else "➡️ "

        print(f"  {arrow} [{i:02d}/{len(images)}] {rel}")
        print(f"        {format_size(b):>10}  →  {format_size(a):<10}  "
              f"(tiết kiệm {format_size(saved)}, -{pct:.0f}%)")

        total_before += b
        total_after  += a

    total_saved = total_before - total_after
    pct_total   = (total_saved / total_before * 100) if total_before > 0 else 0

    print()
    print("=" * 60)
    print("  📊  KẾT QUẢ TỔNG HỢP")
    print("=" * 60)
    print(f"  Dung lượng trước : {format_size(total_before)}")
    print(f"  Dung lượng sau   : {format_size(total_after)}")
    print(f"  Đã tiết kiệm     : {format_size(total_saved)}  (-{pct_total:.1f}%)")
    if BACKUP:
        print(f"\n  💾 Bản gốc được lưu tại: {backup_root}")
    print("\n  ✅ Hoàn thành!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: compress_images.js
/**
 * =============================================================
 *   Script nén ảnh cho Wedding Website
 *   Dùng: Node.js + sharp
 *   Cài : npm install sharp
 *   Chạy: node compress_images.js
 * =============================================================
 *
 *  CẤU HÌNH – chỉnh tại đây nếu cần:
 *  - JPEG_QUALITY : Chất lượng JPG (1-100). Mặc định 75.
 *  - PNG_QUALITY  : Chất lượng PNG (1-100). Mặc định 80.
 *  - MAX_SIZE     : Chiều rộng/cao tối đa (px). Ảnh lớn hơn sẽ thu nhỏ.
 *  - BACKUP       : true = lưu bản gốc vào images/_backup/
 * =============================================================
 */

const sharp = require("sharp");
const fs = require("fs");
const path = require("path");

// ─────────────────────────────────────────────
//  CẤU HÌNH
// ─────────────────────────────────────────────
const IMAGES_DIR   = path.join(__dirname, "images");
const JPEG_QUALITY = 75;     // 1-100
const PNG_QUALITY  = 80;     // 1-100
const MAX_SIZE     = 1920;   // px (cạnh dài nhất)
const BACKUP       = true;   // Lưu bản gốc vào _backup/?
const SKIP_DIRS    = ["_backup"];
// ─────────────────────────────────────────────

const SUPPORTED = [".jpg", ".jpeg", ".png"];

function formatSize(bytes) {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(2)} MB`;
}

function findImages(dir) {
  let results = [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (SKIP_DIRS.includes(entry.name)) continue;
      results = results.concat(findImages(fullPath));
    } else if (entry.isFile()) {
      const ext = path.extname(entry.name).toLowerCase();
      if (SUPPORTED.includes(ext)) results.push(fullPath);
    }
  }
  return results;
}

function ensureDir(dirPath) {
  if (!fs.existsSync(dirPath)) fs.mkdirSync(dirPath, { recursive: true });
}

function backupFile(filePath) {
  const rel = path.relative(IMAGES_DIR, filePath);
  const dest = path.join(IMAGES_DIR, "_backup", rel);
  ensureDir(path.dirname(dest));
  if (!fs.existsSync(dest)) fs.copyFileSync(filePath, dest);
}

async function compressImage(filePath) {
  const sizeBefore = fs.statSync(filePath).size;
  const ext = path.extname(filePath).toLowerCase();

  if (BACKUP) backupFile(filePath);

  // Ghi ra file tạm rồi replace
  const tmpPath = filePath + ".tmp";

  try {
    let pipeline = sharp(filePath).rotate(); // rotate() giữ EXIF orientation

    // Thu nhỏ nếu quá lớn
    pipeline = pipeline.resize(MAX_SIZE, MAX_SIZE, {
      fit: "inside",
      withoutEnlargement: true,
    });

    if (ext === ".png") {
      pipeline = pipeline.png({ quality: PNG_QUALITY, compressionLevel: 9, effort: 10 });
    } else {
      pipeline = pipeline.jpeg({ quality: JPEG_QUALITY, mozjpeg: true });
    }

    await pipeline.toFile(tmpPath);

    const sizeAfter = fs.statSync(tmpPath).size;

    // Chỉ giữ file nén nếu nhỏ hơn bản gốc
    if (sizeAfter < sizeBefore) {
      fs.renameSync(tmpPath, filePath);
      return { sizeBefore, sizeAfter };
    } else {
      fs.unlinkSync(tmpPath);
      return { sizeBefore, sizeAfter: sizeBefore }; // không đổi
    }
  } catch (err) {
    if (fs.existsSync(tmpPath)) fs.unlinkSync(tmpPath);
    throw err;
  }
}

async function main() {
  const LINE = "=".repeat(62);
  const DASH = "-".repeat(62);

  console.log(LINE);
  console.log("  🖼️   SCRIPT NÉN ẢNH – WEDDING WEBSITE");
  console.log(LINE);

  if (!fs.existsSync(IMAGES_DIR)) {
    console.error(`❌ Không tìm thấy thư mục: ${IMAGES_DIR}`);
    process.exit(1);
  }

  const images = findImages(IMAGES_DIR);
  if (images.length === 0) {
    console.log("⚠️  Không tìm thấy ảnh nào.");
    return;
  }

  const backupRoot = path.join(IMAGES_DIR, "_backup");
  console.log(`\n📁 Thư mục ảnh  : ${IMAGES_DIR}`);
  console.log(`🔢 Tổng số ảnh  : ${images.length} file`);
  console.log(`🎯 Chất lượng   : JPG=${JPEG_QUALITY}%, PNG=${PNG_QUALITY}%, Max=${MAX_SIZE}px`);
  if (BACKUP) console.log(`💾 Backup tại   : ${backupRoot}`);
  console.log(`\n${DASH}`);

  let totalBefore = 0;
  let totalAfter  = 0;

  for (let i = 0; i < images.length; i++) {
    const filePath = images[i];
    const rel = path.relative(IMAGES_DIR, filePath);
    const num = `[${String(i + 1).padStart(2, "0")}/${images.length}]`;

    try {
      const { sizeBefore, sizeAfter } = await compressImage(filePath);
      const saved = sizeBefore - sizeAfter;
      const pct   = sizeBefore > 0 ? ((saved / sizeBefore) * 100).toFixed(0) : 0;
      const icon  = saved > 0 ? "✅" : "➡️ ";

      console.log(`  ${icon} ${num} ${rel}`);
      console.log(
        `        ${formatSize(sizeBefore).padStart(10)}  →  ${formatSize(sizeAfter).padEnd(10)}` +
        `  (tiết kiệm ${formatSize(saved)}, -${pct}%)`
      );

      totalBefore += sizeBefore;
      totalAfter  += sizeAfter;
    } catch (err) {
      console.log(`  ⚠️  ${num} ${rel} — Lỗi: ${err.message}`);
      const s = fs.statSync(filePath).size;
      totalBefore += s;
      totalAfter  += s;
    }
  }

  const totalSaved = totalBefore - totalAfter;
  const pctTotal   = totalBefore > 0 ? ((totalSaved / totalBefore) * 100).toFixed(1) : 0;

  console.log(`\n${LINE}`);
  console.log("  📊  KẾT QUẢ TỔNG HỢP");
  console.log(LINE);
  console.log(`  Dung lượng trước : ${formatSize(totalBefore)}`);
  console.log(`  Dung lượng sau   : ${formatSize(totalAfter)}`);
  console.log(`  Đã tiết kiệm     : ${formatSize(totalSaved)}  (-${pctTotal}%)`);
  if (BACKUP) console.log(`\n  💾 Bản gốc được lưu tại: ${backupRoot}`);
  console.log("\n  ✅ Hoàn thành!");
  console.log(LINE);
}

main().catch((err) => {
  console.error("❌ Lỗi nghiêm trọng:", err);
  process.exit(1);
});
